
<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="UTF-8"/>
<meta name="viewport" content="width=device-width, initial-scale=1.0"/>
<title>Contracts · Metacells.jl v0.1.0
</title>
<meta name="title" content="Contracts · Metacells.jl v0.1.0"/>
<meta property="og:title" content="Contracts · Metacells.jl v0.1.0"/>
<meta property="twitter:title" content="Contracts · Metacells.jl v0.1.0"/>
<meta name="description" content="Documentation for Metacells.jl v0.1.0."/>
<meta property="og:description" content="Documentation for Metacells.jl v0.1.0."/>
<meta property="twitter:description" content="Documentation for Metacells.jl v0.1.0."/>
<script data-outdated-warner src="assets/warner.js">
</script>
<link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/>
<script>documenterBaseURL="."
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js">
</script>
<script src="search_index.js">
</script>
<script src="siteinfo.js">
</script>
<script src="../versions.js">
</script>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/>
<script src="assets/themeswap.js">
</script>
</head>
<body>
<div id="documenter">
<nav class="docs-sidebar">
<div class="docs-package-name">
<span class="docs-autofit">
<a href="index.html">Metacells.jl v0.1.0
</a>
</span>
</div>
<button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)
</button>
<ul class="docs-menu">
<li>
<a class="tocitem" href="index.html">Metacells
</a>
</li>
<li>
<a class="tocitem" href="gmara.html">Gmara
</a>
</li>
<li class="is-active">
<a class="tocitem" href="contracts.html">Contracts
</a>
<ul class="internal">
<li>
<a class="tocitem" href="#Axes">
<span>Axes
</span>
</a>
</li>
<li>
<a class="tocitem" href="#Vectors">
<span>Vectors
</span>
</a>
</li>
<li>
<a class="tocitem" href="#Matrices">
<span>Matrices
</span>
</a>
</li>
<li>
<a class="tocitem" href="#Index">
<span>Index
</span>
</a>
</li>
</ul>
</li>
<li>
<a class="tocitem" href="identify_genes.html">Identify Genes
</a>
</li>
<li>
<a class="tocitem" href="collect.html">Collect Metacells
</a>
</li>
<li>
<a class="tocitem" href="blocks.html">Blocks
</a>
</li>
<li>
<a class="tocitem" href="anndata_format.html">AnnData Format
</a>
</li>
<li>
<a class="tocitem" href="defaults.html">Defaults
</a>
</li>
</ul>
<div class="docs-version-selector field has-addons">
<div class="control">
<span class="docs-label button is-static is-size-7">Version
</span>
</div>
<div class="docs-selector control is-expanded">
<div class="select is-fullwidth is-size-7">
<select id="documenter-version-selector">
</select>
</div>
</div>
</div>
</nav>
<div class="docs-main">
<header class="docs-navbar">
<a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#">
</a>
<nav class="breadcrumb">
<ul class="is-hidden-mobile">
<li class="is-active">
<a href="contracts.html">Contracts
</a>
</li>
</ul>
<ul class="is-hidden-tablet">
<li class="is-active">
<a href="contracts.html">Contracts
</a>
</li>
</ul>
</nav>
<div class="docs-right">
<a class="docs-navbar-link" href="https://github.com/tanaylab/Metacells.jl/blob/main{path}?plain=1#L{line}" title="View the repository on GitHub">
<span class="docs-icon fa-brands">
</span>
<span class="docs-label is-hidden-touch">GitHub
</span>
</a>
<a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings">
</a>
<a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings">
</a>
</div>
</header>
<article class="content" id="documenter-page">
<h1 id="Contracts">
<a class="docs-heading-anchor" href="#Contracts">Contracts
</a>
<a id="Contracts-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Contracts" title="Permalink">
</a>
</h1>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts" href="#Metacells.Contracts">
<code>Metacells.Contracts
</code>
</a> — 
<span class="docstring-category">Module
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<p>Functions for defining a 
<code>Contract
</code> for a metacells 
<code>Daf
</code> data set 
<code>@computation
</code>. This also serves as a vocabulary describing the data we keep when doing metacells based analysis, which is a solid basis for understanding what is going on. As Fred Brooks said: &quot;Show me your flowcharts and conceal your tables, and I shall continue to be mystified. Show me your tables, and I won’t usually need your flowcharts; they’ll be obvious.&quot;.
</p>
<p>In the descriptions below, &quot;fold factor&quot; refers to the log base 2 of the ratio between expression levels. For fold factors between gene RNA expression levels (fractions), we typically use a regularization factor of 1e-5 to avoid division by zero.
</p>
</div>
</section>
</article>
<h2 id="Axes">
<a class="docs-heading-anchor" href="#Axes">Axes
</a>
<a id="Axes-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Axes" title="Permalink">
</a>
</h2>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_axis" href="#Metacells.Contracts.gene_axis">
<code>Metacells.Contracts.gene_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of sequenced genes. By convention we use gene symbols as the namespace of the genes, but this may be different depending on the data set.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_axis" href="#Metacells.Contracts.cell_axis">
<code>Metacells.Contracts.cell_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of sequenced single cells. There&#39;s no convention for cell names, as long as they are unique. Typically some sort of barcode is used, possibly combined with a batch and/or plate and/or experiment identification. In the latter case it is recommended that 
<code>batch
</code> and/or 
<code>plate
</code> and/or 
<code>experiment
</code> would also be created as explicit axes, to allow associating metadata with them instead of repeating it for each cell.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_axis" href="#Metacells.Contracts.metacell_axis">
<code>Metacells.Contracts.metacell_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of metacells, which are minimal-sized groups of cells for robust point estimates. Each metacell is considered to be a robustly estimated point in the multi-dimensional manifold of cell states. Metacells may be very similar or very different from each other depending on the data set and the manifold region.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.old_metacell_axis" href="#Metacells.Contracts.old_metacell_axis">
<code>Metacells.Contracts.old_metacell_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">old_metacell_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>Old metacells when computing new ones. This is the same as the regular metacells axis, but renamed to avoid confusion when computing a new set of metacells for some data based on the old metacells.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.new_metacell_axis" href="#Metacells.Contracts.new_metacell_axis">
<code>Metacells.Contracts.new_metacell_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">new_metacell_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>New metacells when improving old ones. This is the same as the regular metacells axis, but renamed to avoid confusion when computing a new set of metacells for some data based on the old metacells.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.type_axis" href="#Metacells.Contracts.type_axis">
<code>Metacells.Contracts.type_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">type_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of types, which are distinct named biological cell states. Types are convenient labels manually assigned to large groups of cells, metacells, blocks, neighborhoods, etc. The resolution of the type labels depends on the data set and the type of analysis. In particular, types are not typically associated with a specific biological cell state, but rather with a set of related biological cell states (possibly along a gradient of such states).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_axis" href="#Metacells.Contracts.block_axis">
<code>Metacells.Contracts.block_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of blocks, which are distinct groups of metacells with &quot;very close&quot; estimated cell state. The metacells in each block all have very close estimates of the skeleton gene expressions (maximal fold factor up to some maximal span).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.principal_component_axis" href="#Metacells.Contracts.principal_component_axis">
<code>Metacells.Contracts.principal_component_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">principal_component_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of principal components of local linear programs.
</p>
</div>
</section>
</article>
<h2 id="Vectors">
<a class="docs-heading-anchor" href="#Vectors">Vectors
</a>
<a id="Vectors-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Vectors" title="Permalink">
</a>
</h2>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_excluded_vector" href="#Metacells.Contracts.gene_is_excluded_vector">
<code>Metacells.Contracts.gene_is_excluded_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_excluded_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that are excluded from consideration. These genes are completely unrelated to the biological behaviors of interest. Not only that, they have strong and variable expression levels; enough to have an global impact on the expression level of the rest of the genes - for example, mitochondrial genes. Such genes make it difficult to estimate the relative expression level of genes between different cell states. Therefore, such genes aren&#39;t even counted in the total UMIs of each cell.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_lateral_vector" href="#Metacells.Contracts.gene_is_lateral_vector">
<code>Metacells.Contracts.gene_is_lateral_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_lateral_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that are lateral to the biological behaviors of interest. These genes may satisfy all criteria for being in a group of cooperating genes, but the biological behavior they participate in isn&#39;t relevant to the behaviors of interest - for example, genes related to cell cycle or stress. Such genes make it harder to focus on the biological behaviors of interest. They are therefore masked out during the analysis.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_marker_vector" href="#Metacells.Contracts.gene_is_marker_vector">
<code>Metacells.Contracts.gene_is_marker_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_marker_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that distinguish between cell states. These genes have a significant expression level at some cell state, as well as a significant range of expression across all cell states, so can be used to distinguish between cell states. Non-marker genes are by definition not useful for such analysis, but marker genes aren&#39;t necessarily useful due to other considerations.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_marker_rank_vector" href="#Metacells.Contracts.gene_marker_rank_vector">
<code>Metacells.Contracts.gene_marker_rank_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_marker_rank_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The relative ranks of the marker genes. The more the gene distinguishes between different cell states, the better (lower) rank it has. That is, 1 is for the &quot;most&quot; marker gene. Non-marker genes are given an extremely high rank (that maximal the data type allows).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_uncorrelated_vector" href="#Metacells.Contracts.gene_is_uncorrelated_vector">
<code>Metacells.Contracts.gene_is_uncorrelated_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_uncorrelated_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that are not correlated with other gene(s). We typically search for groups of genes that act together. Genes that have no correlation with other genes aren&#39;t useful for this sort of analysis, even if they are marker genes.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_covered_vector" href="#Metacells.Contracts.gene_is_covered_vector">
<code>Metacells.Contracts.gene_is_covered_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_covered_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that are covered by the local linear program. When we approximate the manifold using these local linear programs, we only model these genes, ignoring the rest. This allows the approximation to be unaffected by &quot;irrelevant&quot; genes.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_forbidden_vector" href="#Metacells.Contracts.gene_is_forbidden_vector">
<code>Metacells.Contracts.gene_is_forbidden_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_forbidden_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that are forbidden from being used as skeleton factors. When searching for a set of genes that predict the expression of the rest of the genes, we do not consider the forbidden genes.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_skeleton_vector" href="#Metacells.Contracts.gene_is_skeleton_vector">
<code>Metacells.Contracts.gene_is_skeleton_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_skeleton_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that are used to predict the values of the rest of the (covered) genes. The local programs that we use to approximate the manifold are linear combinations of the skeleton genes. A linear combination of a small set of (different) such programs is used to predict the rest of the genes.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_divergence_vector" href="#Metacells.Contracts.gene_divergence_vector">
<code>Metacells.Contracts.gene_divergence_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_divergence_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>For each gene, we scale fold factors of each gene by multiplying with (1 - divergence) of the gene. In particular this is used when considering the distance between gene expressions. Therefore genes with a non-zero divergence will require a higher fold factor
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_is_excluded_vector" href="#Metacells.Contracts.cell_is_excluded_vector">
<code>Metacells.Contracts.cell_is_excluded_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_is_excluded_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of cells that are excluded from consideration. This can be due to any number of reasons - doublets, too low a number of UMIs, to high a percentage of excluded gene UMIs, etc.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_total_UMIs_vector" href="#Metacells.Contracts.cell_total_UMIs_vector">
<code>Metacells.Contracts.cell_total_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_total_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of all the genes in each cell. This doesn&#39;t include the UMIs of genes excluded for any reason.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_total_UMIs_vector" href="#Metacells.Contracts.metacell_total_UMIs_vector">
<code>Metacells.Contracts.metacell_total_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_total_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs used to estimate the fraction of all the genes in each metacell. This is used to estimate the robustness of the gene fraction estimates.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_covered_UMIs_vector" href="#Metacells.Contracts.metacell_covered_UMIs_vector">
<code>Metacells.Contracts.metacell_covered_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_covered_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of the covered genes in each metacell. This is used to estimate the robustness of the gene fraction estimates.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_block_vector" href="#Metacells.Contracts.metacell_block_vector">
<code>Metacells.Contracts.metacell_block_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_block_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The unique block each metacell belongs to. All the metacells in the same block are assumed to be &quot;very close&quot; to each other, so much so that it is excusable to treat them as &quot;the same&quot; cell state.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_type_vector" href="#Metacells.Contracts.metacell_type_vector">
<code>Metacells.Contracts.metacell_type_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_type_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The type each metacell belongs to.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.new_metacell_block_vector" href="#Metacells.Contracts.new_metacell_block_vector">
<code>Metacells.Contracts.new_metacell_block_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">new_metacell_block_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The unique block each new metacell belongs to. This just uses the renamed old metacells axis when computing a new set of metacells for some data based on the old metacells.
</p>
<div class="admonition is-info">
<header class="admonition-header">Note
</header>
<div class="admonition-body">
<p>The old and new metacells are associated with the same set of blocks, at least initially. At some point you would compute brand new blocks for the new metacells, but as these are not based on the old blocks, there&#39;s no need to access both sets of blocks at the same time so no need for &quot;old blocks&quot; and &quot;new blocks&quot;.
</p>
</div>
</div>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.old_metacell_block_vector" href="#Metacells.Contracts.old_metacell_block_vector">
<code>Metacells.Contracts.old_metacell_block_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">old_metacell_block_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The unique block each old metacell belongs to. This just uses the renamed old metacells axis when computing a new set of metacells for some data based on the old metacells.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.type_color_vector" href="#Metacells.Contracts.type_color_vector">
<code>Metacells.Contracts.type_color_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">type_color_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A unique color for each type for graphs.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_n_principal_components_vector" href="#Metacells.Contracts.block_n_principal_components_vector">
<code>Metacells.Contracts.block_n_principal_components_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_n_principal_components_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The number of used principal components of the local linear model for each block. Since 
<code>Daf
</code> can only deal with fixed-sized axes, we create a &quot;too large&quot; principal components axis. Each block actually uses a smaller (different) number of principal components that hopefully reflects the actual dimensionality of the manifold at that point. The coefficients of the rest of the principal components are set to zero.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_linear_RMSE_vector" href="#Metacells.Contracts.block_linear_RMSE_vector">
<code>Metacells.Contracts.block_linear_RMSE_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_linear_RMSE_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The root mean squared error of predicting the covered genes using the local linear models. This model computes principal components from the skeleton genes and then uses these components to predict the values of all covered genes.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_linear_XRMSE_vector" href="#Metacells.Contracts.block_linear_XRMSE_vector">
<code>Metacells.Contracts.block_linear_XRMSE_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_linear_XRMSE_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The cross-validated root mean squared error of predicting the covered genes using the linear model. This is similar to the RMSE, but is computed using cross-validation (training a linear model on a subset of the local metacells and evaluating it on the rest). This will be higher than the RMSE; the difference is an indicator of the &quot;overfitting&quot; of the linear model, which should be low relative to the total RMSE.
</p>
</div>
</section>
</article>
<h2 id="Matrices">
<a class="docs-heading-anchor" href="#Matrices">Matrices
</a>
<a id="Matrices-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Matrices" title="Permalink">
</a>
</h2>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_gene_UMIs_matrix" href="#Metacells.Contracts.cell_gene_UMIs_matrix">
<code>Metacells.Contracts.cell_gene_UMIs_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_gene_UMIs_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The number of UMIs collected for each gene for each cell. This is the &quot;ground truth&quot; everything else is built on. The total number of UMIs per cell differs (sometimes wildly) based on the scRNA-seq technology and protocol.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_metacell_vector" href="#Metacells.Contracts.cell_metacell_vector">
<code>Metacells.Contracts.cell_metacell_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_metacell_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The unique metacell each cell belongs to. All the cells in the same metacell are assumed to have &quot;the same&quot; (relevant) biological state. This is the empty string if the cell does not belong to any metacell (is excluded or outlier).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_new_metacell_vector" href="#Metacells.Contracts.cell_new_metacell_vector">
<code>Metacells.Contracts.cell_new_metacell_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_new_metacell_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The unique new metacell each cell belongs to. This just uses the renamed new metacells axis when computing a new set of metacells for some data based on the old metacells.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_old_metacell_vector" href="#Metacells.Contracts.cell_old_metacell_vector">
<code>Metacells.Contracts.cell_old_metacell_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_old_metacell_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The unique old metacell each cell belongs to. This just uses the renamed old metacells axis when computing a new set of metacells for some data based on the old metacells.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_gene_fraction_matrix" href="#Metacells.Contracts.metacell_gene_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_gene_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The estimated fraction of the UMIs of each gene in each metacell. We assume that each metacell is a sample of the manifold, representing a real biological state, regardless to its distance to other metacells (subject to cleaning up batch effects, purging doublets, and compensating for any other technical artifacts).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_gene_total_UMIs_matrix" href="#Metacells.Contracts.metacell_gene_total_UMIs_matrix">
<code>Metacells.Contracts.metacell_gene_total_UMIs_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_gene_total_UMIs_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs used to estimate the fraction of each gene in each metacell. This is used to estimate the robustness of the estimate. When computing fold factors, we require the total number of UMIs (from both compared estimates) to be some minimum, and possibly adjust the fold factor according to some confidence level (assuming a multinomial sampling distribution).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_gene_covered_fraction_matrix" href="#Metacells.Contracts.metacell_gene_covered_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_covered_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_gene_covered_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The relative fraction of covered genes. This ensures the sum of the fractions of the covered genes is one in each metacell, to ignore the impact of any residual uncovered gene programs.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_gene_scaled_log_covered_fraction_matrix" href="#Metacells.Contracts.metacell_gene_scaled_log_covered_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_scaled_log_covered_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_gene_scaled_log_covered_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The log of the relative fraction of covered genes, scaled by divergence. This computes the log base 2 of the 
<code>metacell_gene_covered_fraction_matrix
</code>, with some regularization, and then scales the result using the 
<a href="contracts.html#Metacells.Contracts.gene_divergence_vector">
<code>gene_divergence_vector
</code>
</a> so that the range of expression of &quot;wild&quot; genes is reduced. The end result is what we actually try to fit with local linear programs.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.old_metacell_gene_scaled_log_covered_fraction_matrix" href="#Metacells.Contracts.old_metacell_gene_scaled_log_covered_fraction_matrix">
<code>Metacells.Contracts.old_metacell_gene_scaled_log_covered_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">old_metacell_gene_scaled_log_covered_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The log of the relative fraction of covered genes, scaled by divergence. This just uses the renamed old metacells axis when computing a new set of metacells for some data based on the old metacells.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_block_is_in_neighborhood_matrix" href="#Metacells.Contracts.block_block_is_in_neighborhood_matrix">
<code>Metacells.Contracts.block_block_is_in_neighborhood_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_block_is_in_neighborhood_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>For each block, the mask of nearby blocks in its immediate neighborhood. The neighborhood consists of a small number of very close blocks, which we use as the basis for evaluating linear approximations (of the log of the expression of the genes).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_block_is_in_environment_matrix" href="#Metacells.Contracts.block_block_is_in_environment_matrix">
<code>Metacells.Contracts.block_block_is_in_environment_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_block_is_in_environment_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>For each block, the mask of nearby blocks in its linear environment. We expect the covered genes in the metacells in the blocks of the environment to reasonably fit a local linear model (of the log of the expression).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_fraction_matrix" href="#Metacells.Contracts.block_gene_fraction_matrix">
<code>Metacells.Contracts.block_gene_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The estimated fraction of the UMIs of each gene in each block. This assumes the metacells in each block are samples of the &quot;same&quot; biological state, based on the fact that all these metacells have very similar expression levels for all the global predictive transcription factors.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_total_UMIs_matrix" href="#Metacells.Contracts.block_gene_total_UMIs_matrix">
<code>Metacells.Contracts.block_gene_total_UMIs_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_total_UMIs_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs used to estimate the fraction of each gene in each block. This is used to estimate the robustness of the estimate. When computing fold factors, we require the total number of UMIs (from both compared estimates) to be some minimum, and possibly adjust the fold factor according to some confidence level (assuming a multinomial sampling distribution).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_mean_scaled_log_covered_fraction_matrix" href="#Metacells.Contracts.block_gene_mean_scaled_log_covered_fraction_matrix">
<code>Metacells.Contracts.block_gene_mean_scaled_log_covered_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_mean_scaled_log_covered_fraction_matrix(
    expectation::ContractExpectation,
)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The mean in the metacells of the block of the log of the relative fraction of covered genes, scaled by divergence. This is used by the linear model; specifically, the coefficients are applied to the difference from this mean, and the results are added back to this mean. This is zero for uncovered genes.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_principal_component_is_used_matrix" href="#Metacells.Contracts.block_principal_component_is_used_matrix">
<code>Metacells.Contracts.block_principal_component_is_used_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_principal_component_is_used_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>Whether each principal component is used by each block. This mask is 
<code>true
</code> for the first used principal components of each block and 
<code>false
</code> for the rest. It is useful for constructing 
<code>Daf
</code> queries that only return coefficients for the used principal components for a specific block (e.g., 
<code>gene &amp; is_skeleton / principal_component &amp; is_used ; block = $(block_name) : $(block_name)_skeleton_coefficient
</code>).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_principal_component_gene_covered_coefficient_tensor" href="#Metacells.Contracts.block_principal_component_gene_covered_coefficient_tensor">
<code>Metacells.Contracts.block_principal_component_gene_covered_coefficient_tensor
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_principal_component_gene_covered_coefficient_tensor(
    expectation::ContractExpectation
)::Pair{TensorKey, DataSpecification}
</code>
</pre>
<p>The coefficient of each principal component for computing each covered gene. This is zero for uncovered genes. Combined with the skeleton genes coefficients this gives a local linear model for estimating all the covered genes (
<em>including
</em> the skeleton genes).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_principal_component_gene_skeleton_coefficient_tensor" href="#Metacells.Contracts.block_principal_component_gene_skeleton_coefficient_tensor">
<code>Metacells.Contracts.block_principal_component_gene_skeleton_coefficient_tensor
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_principal_component_gene_skeleton_coefficient_tensor(
    expectation::ContractExpectation
)::Pair{TensorKey, DataSpecification}
</code>
</pre>
<p>The coefficient of each skeleton gene for computing a principle component. This is zero for non-skeleton genes. The coefficients can be either positive or negative. Combined with the covered genes coefficients this gives a local linear model for estimating all the covered genes.
</p>
</div>
</section>
</article>
<h2 id="Index">
<a class="docs-heading-anchor" href="#Index">Index
</a>
<a id="Index-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Index" title="Permalink">
</a>
</h2>
<ul>
<li>
<a href="contracts.html#Metacells.Contracts">
<code>Metacells.Contracts
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_axis">
<code>Metacells.Contracts.block_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_block_is_in_environment_matrix">
<code>Metacells.Contracts.block_block_is_in_environment_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_block_is_in_neighborhood_matrix">
<code>Metacells.Contracts.block_block_is_in_neighborhood_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_fraction_matrix">
<code>Metacells.Contracts.block_gene_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_mean_scaled_log_covered_fraction_matrix">
<code>Metacells.Contracts.block_gene_mean_scaled_log_covered_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_total_UMIs_matrix">
<code>Metacells.Contracts.block_gene_total_UMIs_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_linear_RMSE_vector">
<code>Metacells.Contracts.block_linear_RMSE_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_linear_XRMSE_vector">
<code>Metacells.Contracts.block_linear_XRMSE_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_n_principal_components_vector">
<code>Metacells.Contracts.block_n_principal_components_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_principal_component_gene_covered_coefficient_tensor">
<code>Metacells.Contracts.block_principal_component_gene_covered_coefficient_tensor
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_principal_component_gene_skeleton_coefficient_tensor">
<code>Metacells.Contracts.block_principal_component_gene_skeleton_coefficient_tensor
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_principal_component_is_used_matrix">
<code>Metacells.Contracts.block_principal_component_is_used_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_axis">
<code>Metacells.Contracts.cell_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_gene_UMIs_matrix">
<code>Metacells.Contracts.cell_gene_UMIs_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_is_excluded_vector">
<code>Metacells.Contracts.cell_is_excluded_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_metacell_vector">
<code>Metacells.Contracts.cell_metacell_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_new_metacell_vector">
<code>Metacells.Contracts.cell_new_metacell_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_old_metacell_vector">
<code>Metacells.Contracts.cell_old_metacell_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_total_UMIs_vector">
<code>Metacells.Contracts.cell_total_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_axis">
<code>Metacells.Contracts.gene_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_divergence_vector">
<code>Metacells.Contracts.gene_divergence_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_covered_vector">
<code>Metacells.Contracts.gene_is_covered_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_excluded_vector">
<code>Metacells.Contracts.gene_is_excluded_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_forbidden_vector">
<code>Metacells.Contracts.gene_is_forbidden_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_lateral_vector">
<code>Metacells.Contracts.gene_is_lateral_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_marker_vector">
<code>Metacells.Contracts.gene_is_marker_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_skeleton_vector">
<code>Metacells.Contracts.gene_is_skeleton_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_uncorrelated_vector">
<code>Metacells.Contracts.gene_is_uncorrelated_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_marker_rank_vector">
<code>Metacells.Contracts.gene_marker_rank_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_axis">
<code>Metacells.Contracts.metacell_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_block_vector">
<code>Metacells.Contracts.metacell_block_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_covered_UMIs_vector">
<code>Metacells.Contracts.metacell_covered_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_gene_covered_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_covered_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_gene_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_gene_scaled_log_covered_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_scaled_log_covered_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_gene_total_UMIs_matrix">
<code>Metacells.Contracts.metacell_gene_total_UMIs_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_total_UMIs_vector">
<code>Metacells.Contracts.metacell_total_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_type_vector">
<code>Metacells.Contracts.metacell_type_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.new_metacell_axis">
<code>Metacells.Contracts.new_metacell_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.new_metacell_block_vector">
<code>Metacells.Contracts.new_metacell_block_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.old_metacell_axis">
<code>Metacells.Contracts.old_metacell_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.old_metacell_block_vector">
<code>Metacells.Contracts.old_metacell_block_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.old_metacell_gene_scaled_log_covered_fraction_matrix">
<code>Metacells.Contracts.old_metacell_gene_scaled_log_covered_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.principal_component_axis">
<code>Metacells.Contracts.principal_component_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.type_axis">
<code>Metacells.Contracts.type_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.type_color_vector">
<code>Metacells.Contracts.type_color_vector
</code>
</a>
</li>
</ul>
</article>
<nav class="docs-footer">
<a class="docs-footer-prevpage" href="gmara.html">« Gmara
</a>
<a class="docs-footer-nextpage" href="identify_genes.html">Identify Genes »
</a>
<div class="flexbox-break">
</div>
<p class="footer-message">Powered by 
<a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl
</a> and the 
<a href="https://julialang.org/">Julia Programming Language
</a>.
</p>
</nav>
</div>
<div class="modal" id="documenter-settings">
<div class="modal-background">
</div>
<div class="modal-card">
<header class="modal-card-head">
<p class="modal-card-title">Settings
</p>
<button class="delete">
</button>
</header>
<section class="modal-card-body">
<p>
<label class="label">Theme
</label>
<div class="select">
<select id="documenter-themepicker">
<option value="auto">Automatic (OS)
</option>
<option value="documenter-light">documenter-light
</option>
<option value="documenter-dark">documenter-dark
</option>
<option value="catppuccin-latte">catppuccin-latte
</option>
<option value="catppuccin-frappe">catppuccin-frappe
</option>
<option value="catppuccin-macchiato">catppuccin-macchiato
</option>
<option value="catppuccin-mocha">catppuccin-mocha
</option>
</select>
</div>
</p>
<hr/>
<p>This document was generated with 
<a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl
</a> version 1.10.1. Using Julia version 1.11.4.
</p>
</section>
<footer class="modal-card-foot">
</footer>
</div>
</div>
</div>
</body>
</html>
