
<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="UTF-8"/>
<meta name="viewport" content="width=device-width, initial-scale=1.0"/>
<title>Contracts · Metacells.jl v0.1.0
</title>
<meta name="title" content="Contracts · Metacells.jl v0.1.0"/>
<meta property="og:title" content="Contracts · Metacells.jl v0.1.0"/>
<meta property="twitter:title" content="Contracts · Metacells.jl v0.1.0"/>
<meta name="description" content="Documentation for Metacells.jl v0.1.0."/>
<meta property="og:description" content="Documentation for Metacells.jl v0.1.0."/>
<meta property="twitter:description" content="Documentation for Metacells.jl v0.1.0."/>
<script data-outdated-warner src="assets/warner.js">
</script>
<link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/>
<script>documenterBaseURL="."
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js">
</script>
<script src="search_index.js">
</script>
<script src="siteinfo.js">
</script>
<script src="../versions.js">
</script>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/>
<script src="assets/themeswap.js">
</script>
</head>
<body>
<div id="documenter">
<nav class="docs-sidebar">
<div class="docs-package-name">
<span class="docs-autofit">
<a href="index.html">Metacells.jl v0.1.0
</a>
</span>
</div>
<button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)
</button>
<ul class="docs-menu">
<li>
<a class="tocitem" href="index.html">Metacells
</a>
</li>
<li>
<a class="tocitem" href="gmara.html">Gmara
</a>
</li>
<li class="is-active">
<a class="tocitem" href="contracts.html">Contracts
</a>
<ul class="internal">
<li>
<a class="tocitem" href="#Genes">
<span>Genes
</span>
</a>
</li>
<li>
<a class="tocitem" href="#Cells">
<span>Cells
</span>
</a>
</li>
<li>
<a class="tocitem" href="#Metacells">
<span>Metacells
</span>
</a>
</li>
<li>
<a class="tocitem" href="#Blocks">
<span>Blocks
</span>
</a>
</li>
<li>
<a class="tocitem" href="#Type-Annotations">
<span>Type Annotations
</span>
</a>
</li>
<li>
<a class="tocitem" href="#Modules">
<span>Modules
</span>
</a>
</li>
<li>
<a class="tocitem" href="#Index">
<span>Index
</span>
</a>
</li>
</ul>
</li>
<li>
<a class="tocitem" href="analyze_cells.html">Analyze Cells
</a>
</li>
<li>
<a class="tocitem" href="analyze_metacells.html">Analyze Metacells
</a>
</li>
<li>
<a class="tocitem" href="compute_blocks.html">Compute Blocks
</a>
</li>
<li>
<a class="tocitem" href="analyze_blocks.html">Analyze Blocks
</a>
</li>
<li>
<a class="tocitem" href="compute_modules.html">Compute Modules
</a>
</li>
<li>
<a class="tocitem" href="analyze_modules.html">Analyze Modules
</a>
</li>
<li>
<a class="tocitem" href="sharpen_metacells.html">Sharpen Metacells
</a>
</li>
<li>
<a class="tocitem" href="anndata_format.html">AnnData Format
</a>
</li>
<li>
<a class="tocitem" href="defaults.html">Defaults
</a>
</li>
</ul>
<div class="docs-version-selector field has-addons">
<div class="control">
<span class="docs-label button is-static is-size-7">Version
</span>
</div>
<div class="docs-selector control is-expanded">
<div class="select is-fullwidth is-size-7">
<select id="documenter-version-selector">
</select>
</div>
</div>
</div>
</nav>
<div class="docs-main">
<header class="docs-navbar">
<a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#">
</a>
<nav class="breadcrumb">
<ul class="is-hidden-mobile">
<li class="is-active">
<a href="contracts.html">Contracts
</a>
</li>
</ul>
<ul class="is-hidden-tablet">
<li class="is-active">
<a href="contracts.html">Contracts
</a>
</li>
</ul>
</nav>
<div class="docs-right">
<a class="docs-navbar-link" href="https://github.com/tanaylab/Metacells.jl/blob/main{path}?plain=1#L{line}" title="View the repository on GitHub">
<span class="docs-icon fa-brands">
</span>
<span class="docs-label is-hidden-touch">GitHub
</span>
</a>
<a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings">
</a>
<a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings">
</a>
</div>
</header>
<article class="content" id="documenter-page">
<h1 id="Contracts">
<a class="docs-heading-anchor" href="#Contracts">Contracts
</a>
<a id="Contracts-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Contracts" title="Permalink">
</a>
</h1>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts" href="#Metacells.Contracts">
<code>Metacells.Contracts
</code>
</a> — 
<span class="docstring-category">Module
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<p>Functions for defining a 
<code>Contract
</code> for a metacells 
<code>Daf
</code> data set 
<code>@computation
</code>. This also serves as a vocabulary describing the data we keep when doing metacells based analysis, which is a solid basis for understanding what is going on. As Fred Brooks said: &quot;Show me your flowcharts and conceal your tables, and I shall continue to be mystified. Show me your tables, and I won’t usually need your flowcharts; they’ll be obvious.&quot;.
</p>
<p>We often look at the log base 2 of the gene expression level. We use a regularization factor of 1e-5 to avoid a log of zero. This means the minimal log value is -16.6 and the maximal value for a well-behaved gene (~0.1% of the total UMIs of the cell) is -10. We typically look at the range -16 &quot;basically no expression&quot; to -10 &quot;respectably high expression&quot;.
</p>
<p>In the descriptions below, &quot;fold factor&quot; refers to the log base 2 of the ratio between expression levels. That is, we look at an interesting fold factor of ~6 between &quot;basically no expression&quot; and &quot;respectably high expression&quot;. A fold factor (difference in logs) of ~3 (that is, ~8x expression level) is typically taken to be &quot;significant&quot; for a single gene between cells by the metacell algorithm. However lower fold factors may be significant, especially as the number of UMIs used to estimate the expression levels are larger (e.g., between expression levels of metacells). As a quick and useful heuristic, we don&#39;t consider fold factors between metacell genes to be meaningful unless the sum of the gene UMIs in both compared metacells is too low (below 40).
</p>
<p>Genes that are higher than -10 (that is, more than 0.1%) are &quot;bursty&quot;, e.g. Hemoglobins and Cytokines. Sometimes such genes take more than 10% of the total UMIs of a cell! TODO: Special handling for such genes, as they effect the denominator for linear fraction estimations and overwhelm least-square approximations.
</p>
</div>
</section>
</article>
<h2 id="Genes">
<a class="docs-heading-anchor" href="#Genes">Genes
</a>
<a id="Genes-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Genes" title="Permalink">
</a>
</h2>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_axis" href="#Metacells.Contracts.gene_axis">
<code>Metacells.Contracts.gene_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of sequenced genes. By convention we use gene symbols as the namespace of the genes, but this may be different depending on the data set.
</p>
</div>
</section>
</article>
<h3 id="Genes-Masks">
<a class="docs-heading-anchor" href="#Genes-Masks">Genes Masks
</a>
<a id="Genes-Masks-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Genes-Masks" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_mitochondrial_vector" href="#Metacells.Contracts.gene_is_mitochondrial_vector">
<code>Metacells.Contracts.gene_is_mitochondrial_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_mitochondrial_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of mitochondrial genes. These genes are typically excluded from the analysis as they are (mostly) unrelated to the biological behaviors of interest, and also have such a large and variable expression level that including them would skew the denominator when estimating linear gene expression level fractions.
</p>
<p>This vector is created in a supervised way based on biological and technical considerations.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_ribosomal_vector" href="#Metacells.Contracts.gene_is_ribosomal_vector">
<code>Metacells.Contracts.gene_is_ribosomal_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_ribosomal_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of mitochondrial genes. These genes are typically excluded from the analysis, as they aren&#39;t always unrelated to the biological behaviors of interest, they have such a large and variable expression level that including them would skew the denominator when estimating linear gene expression level fractions.
</p>
<p>This vector is created in a supervised way based on biological and technical considerations.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_excluded_vector" href="#Metacells.Contracts.gene_is_excluded_vector">
<code>Metacells.Contracts.gene_is_excluded_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_excluded_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that are excluded from consideration. These genes are either unrelated to the biological behaviors of interest and/or have such a large and variable expression level that including them would skew the denominator when estimating linear genes expression level fractions.
</p>
<p>This vector is created in a supervised way based on biological and technical considerations.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_lateral_vector" href="#Metacells.Contracts.gene_is_lateral_vector">
<code>Metacells.Contracts.gene_is_lateral_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_lateral_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that are lateral to the biological behaviors of interest. These genes may satisfy all criteria for being in a group of cooperating genes, but the biological behavior they participate in isn&#39;t relevant to the behaviors of interest - for example, genes related to cell cycle or stress. Such genes make it harder to focus on the biological behaviors of interest. They are therefore masked out during some phases of the analysis. However, unlike excluded genes, we still track their expression level and take it into account (e.g., when removing outlier cells from metacells).
</p>
<p>This vector is created in a supervised way based on biological considerations.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_marker_vector" href="#Metacells.Contracts.gene_is_marker_vector">
<code>Metacells.Contracts.gene_is_marker_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_marker_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that distinguish between cell states. These genes have a significant expression level at some cell state, as well as a significant range of expression across all cell states, so can be used to distinguish between cell states. Non-marker genes are by definition not useful for such analysis, but marker genes aren&#39;t necessarily useful due to other considerations (e.g., they may be lateral genes).
</p>
<p>This vector is populated by 
<a href="analyze_genes.html#Metacells.AnalyzeGenes.identify_marker_genes!">
<code>identify_marker_genes!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_marker_rank_vector" href="#Metacells.Contracts.gene_marker_rank_vector">
<code>Metacells.Contracts.gene_marker_rank_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_marker_rank_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The relative ranks of the marker genes. The more the gene distinguishes between different cell states, the better (lower) rank it has. That is, 1 is the &quot;most&quot; marker gene. Non-marker genes are given an extremely high rank (that maximal the data type allows).
</p>
<p>This vector is populated by 
<a href="analyze_genes.html#Metacells.AnalyzeGenes.rank_marker_genes!">
<code>rank_marker_genes!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_covered_vector" href="#Metacells.Contracts.gene_is_covered_vector">
<code>Metacells.Contracts.gene_is_covered_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_covered_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that are covered by the model(s). When we construct some model(s) for approximating the manifold describing the cell behaviors, we focus only on these genes, ignoring the rest. This allows the model(s) to be more sensitive to variations in the covered genes, without the distraction of the noise from the rest. The covered genes are typically some subset of the marker genes; in particular, they don&#39;t include lateral genes. Properties that are computed using only the covered genes contain 
<code>_covered_
</code> in their name, e.g., 
<code>covered_fraction
</code> vs. 
<code>linear_fraction
</code>.
</p>
<p>This vector is populated by 
<a href="analyze_genes.html#Metacells.AnalyzeGenes.identify_covered_genes!">
<code>identify_covered_genes!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_regulator_vector" href="#Metacells.Contracts.gene_is_regulator_vector">
<code>Metacells.Contracts.gene_is_regulator_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_regulator_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that regulate the expression level of other genes. These genes are expected to be at the core of the gene programs that describe the cell behaviors manifold.
</p>
<p>This vector is populated by 
<a href="analyze_genes.html#Metacells.AnalyzeGenes.fetch_regulators!">
<code>fetch_regulators!
</code>
</a>.
</p>
<div class="admonition is-info" id="Note-728f1ce25f8644c5">
<header class="admonition-header">Note
<a class="admonition-anchor" href="#Note-728f1ce25f8644c5" title="Permalink">
</a>
</header>
<div class="admonition-body">
<p>These are 
<strong>not
</strong> all the genes that bind to DNA (aka &quot;transcription factors&quot;).
</p>
</div>
</div>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_skeleton_vector" href="#Metacells.Contracts.gene_is_skeleton_vector">
<code>Metacells.Contracts.gene_is_skeleton_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_skeleton_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that are used to predict the values of the rest of the (covered) genes. We assume that knowing the expression level of the skeleton genes is sufficient to reasonably estimate the expression level of the rest of the covered genes. For example, if two metacells have &quot;very close&quot; expression level of all the skeleton genes, we assume that these metacells are &quot;very similar&quot;, without looking at the rest of the covered genes.
</p>
<p>This vector is populated by 
<a href="analyze_genes.html#Metacells.AnalyzeGenes.identify_skeleton_genes!">
<code>identify_skeleton_genes!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.gene_is_forbidden_vector" href="#Metacells.Contracts.gene_is_forbidden_vector">
<code>Metacells.Contracts.gene_is_forbidden_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">gene_is_forbidden_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of genes that are forbidden from being used as skeleton genes.
</p>
<p>This vector is created in a supervised way based on biological and technical considerations.
</p>
</div>
</section>
</article>
<h2 id="Cells">
<a class="docs-heading-anchor" href="#Cells">Cells
</a>
<a id="Cells-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Cells" title="Permalink">
</a>
</h2>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_axis" href="#Metacells.Contracts.cell_axis">
<code>Metacells.Contracts.cell_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of sequenced single cells. There&#39;s no convention for cell names, as long as they are unique. Typically some sort of barcode is used, possibly combined with a batch and/or plate and/or experiment identification. In the latter case it is recommended that 
<code>batch
</code> and/or 
<code>plate
</code> and/or 
<code>experiment
</code> would also be created as explicit axes, to allow associating metadata with them instead of repeating it for each cell.
</p>
</div>
</section>
</article>
<h3 id="Cells-UMIs">
<a class="docs-heading-anchor" href="#Cells-UMIs">Cells UMIs
</a>
<a id="Cells-UMIs-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Cells-UMIs" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_gene_UMIs_matrix" href="#Metacells.Contracts.cell_gene_UMIs_matrix">
<code>Metacells.Contracts.cell_gene_UMIs_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_gene_UMIs_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The number of UMIs collected for each gene for each cell. This is the measurement everything else is built on. This is some sampling of the ground truth of the number of RNA molecules of each gene in each cell. This suffers from randomness (only a small subset of the molecules are actually counted), and biases (some molecules are favored over others, depending on the specific technology and protocol used). More fundamentally, the total number of RNA molecules varies (sometimes wildly) between different cells.
</p>
<p>The UMIs count is therefore very sparse and very noisy. The challenge of scRNA-seq analysis is to extract meaningful patterns (cooperating gene programs) that describe the cell behaviors based on this data.
</p>
<p>To analyze this data, we are forced to normalize the UMIs counts to a fraction of the total. This is mostly meaningless in the single-cell level because we have so few UMIs per each cell (x,000 or x0,000 UMIs out of the x00,000 RNA molecules) and low-expressing (but still important) genes may have one or at most few UMIs.
</p>
<p>Therefore fractions are only computed for collection of cells. Metacells are defined as the smallest such collections that can be used to give meaningful fraction estimates (therefore maximizing sensitivity for capturing all cell behaviors). This is in contrast with classical methods that look at very large collections (cell &quot;types&quot;).
</p>
<p>Using fractions is sensitive to the choice of the denominator; a gene (or gene program) with very high expression would artificially reduce the fraction of the rest of the genes. We therefore exclude known troublesome genes from the denominator (e.g., mitochondrial and ribosomal genes). When constructing parametric models, we recompute the fractions to just the covered genes, to reduce the impact on the denominator of lateral genes (e.g., cell cycle). And when comparing fractions between different data sets, one must 
<em>always
</em> renormalize the fractions to the common genes set.
</p>
<p>Ideally, we could somehow create an estimate of the concentration of each RNA molecule as this is in absolute units so is directly comparable across analysis methods and data sets. However, in addition to being dependent on the unknown total number of molecules in each cell, this also depends on the unknown and cell volume.
</p>
<p>This data is obtained from scRNA-seq experiments.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_mitochondrial_UMIs_vector" href="#Metacells.Contracts.cell_mitochondrial_UMIs_vector">
<code>Metacells.Contracts.cell_mitochondrial_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_mitochondrial_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of all the mitochondrial genes in each cell.
</p>
<p>This vector is created in a supervised way, typically based on the naming convention that mitochondrial genes names match the pattern 
<code>^MT-.*$
</code>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_ribosomal_UMIs_vector" href="#Metacells.Contracts.cell_ribosomal_UMIs_vector">
<code>Metacells.Contracts.cell_ribosomal_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_ribosomal_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of all the ribosomal genes in each cell.
</p>
<p>This vector is created in a supervised way, typically based on the naming convention that mitochondrial genes names match the pattern 
<code>^RP[LS]-.*$
</code>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_excluded_UMIs_vector" href="#Metacells.Contracts.cell_excluded_UMIs_vector">
<code>Metacells.Contracts.cell_excluded_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_excluded_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of all the excluded genes in each cell.
</p>
<p>This vector is created in a supervised way based on biological and technical considerations. It typically includes mitochondrial and ribosomal genes, as well as other genes which are highly correlated with other excluded genes.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_total_UMIs_vector" href="#Metacells.Contracts.cell_total_UMIs_vector">
<code>Metacells.Contracts.cell_total_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_total_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of all the non-excluded genes in each cell.
</p>
<p>This vector is populated by 
<a href="analyze_cells.html#Metacells.AnalyzeCells.compute_cells_total_UMIs!">
<code>compute_cells_total_UMIs!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_covered_UMIs_vector" href="#Metacells.Contracts.cell_covered_UMIs_vector">
<code>Metacells.Contracts.cell_covered_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_covered_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of all the covered genes in each cell.
</p>
<p>This vector is populated by 
<a href="analyze_cells.html#Metacells.AnalyzeCells.compute_cells_covered_UMIs!">
<code>compute_cells_covered_UMIs!
</code>
</a>.
</p>
</div>
</section>
</article>
<h3 id="Cells-masks">
<a class="docs-heading-anchor" href="#Cells-masks">Cells masks
</a>
<a id="Cells-masks-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Cells-masks" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_is_excluded_vector" href="#Metacells.Contracts.cell_is_excluded_vector">
<code>Metacells.Contracts.cell_is_excluded_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_is_excluded_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A mask of cells that are excluded from consideration. This can be due to any number of reasons - doublets, too low a number of UMIs, to high a percentage of excluded gene UMIs, etc. Excluded cells are not grouped into metacells etc., so they are ignored in all properties that sum UMIs per group (e.g., 
<code>total_UMIs
</code> per metacell).
</p>
<p>This vector is created in a supervised way based on biological and technical considerations.
</p>
</div>
</section>
</article>
<h2 id="Metacells">
<a class="docs-heading-anchor" href="#Metacells">Metacells
</a>
<a id="Metacells-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Metacells" title="Permalink">
</a>
</h2>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_axis" href="#Metacells.Contracts.metacell_axis">
<code>Metacells.Contracts.metacell_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of metacells, which are minimal-sized groups of cells for robust fraction estimates. Each metacell is considered to be a robustly estimated point in the multi-dimensional manifold of cell states. Metacells may be very similar or very different from each other depending on how well a region of the manifold was sampled; that is, highly sampled cell states would have a large number of higher quality metacells that are more similar (dare one say, even &quot;identical&quot;) to each other, while sparsely sampled cell states would have a smaller number of lower quality metacells that are more different from one another. At the threshold of the sensitivity of the algorithm, very rare cell states would be discarded as outliers, or not even appear in the data in the first place.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_metacell_vector" href="#Metacells.Contracts.cell_metacell_vector">
<code>Metacells.Contracts.cell_metacell_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_metacell_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The unique metacell each cell belongs to. All the cells in the same metacell are assumed to have &quot;the same&quot; (relevant) biological state. This is the empty string if the cell does not belong to any metacell (is excluded or outlier).
</p>
<p>This vector can be populated by any metacells-like algorithm (the original R metacells (1) algorithm, the Python metacells (2) algorithm, other similar algorithms).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_n_cells_vector" href="#Metacells.Contracts.metacell_n_cells_vector">
<code>Metacells.Contracts.metacell_n_cells_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_n_cells_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The number of cells in each metacell.
</p>
<p>This vector is populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_n_cells!">
<code>compute_metacells_n_cells!
</code>
</a>.
</p>
</div>
</section>
</article>
<h3 id="Metacells-UMIs">
<a class="docs-heading-anchor" href="#Metacells-UMIs">Metacells UMIs
</a>
<a id="Metacells-UMIs-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Metacells-UMIs" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_gene_UMIs_matrix" href="#Metacells.Contracts.metacell_gene_UMIs_matrix">
<code>Metacells.Contracts.metacell_gene_UMIs_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_gene_UMIs_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of each gene in the cells of each metacell.
</p>
<p>This matrix is populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_genes_UMIs!">
<code>compute_metacells_genes_UMIs!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_total_UMIs_vector" href="#Metacells.Contracts.metacell_total_UMIs_vector">
<code>Metacells.Contracts.metacell_total_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_total_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of all the non-excluded genes in each metacell. This is used as the denominator for 
<code>linear_fraction
</code>.
</p>
<p>This vector is populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_total_UMIs!">
<code>compute_metacells_total_UMIs!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_covered_UMIs_vector" href="#Metacells.Contracts.metacell_covered_UMIs_vector">
<code>Metacells.Contracts.metacell_covered_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_covered_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of the covered genes in each metacell. This is used as the denominator for 
<code>covered_fraction
</code>.
</p>
<p>This vector is populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_covered_UMIs!">
<code>compute_metacells_covered_UMIs!
</code>
</a>.
</p>
</div>
</section>
</article>
<h3 id="Metacells-Fractions">
<a class="docs-heading-anchor" href="#Metacells-Fractions">Metacells Fractions
</a>
<a id="Metacells-Fractions-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Metacells-Fractions" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_gene_linear_fraction_matrix" href="#Metacells.Contracts.metacell_gene_linear_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_linear_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_gene_linear_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The linear fraction of the UMIs of each non-excluded gene in each metacell, out of the total UMIs. This is the &quot;best&quot; estimate assuming unbiased multinomial sampling. However, this is sensitive to a few cells with very high expression levels (&quot;bursty&quot; genes), as these impact the denominator.
</p>
<p>This matrix is populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_genes_linear_fractions!">
<code>compute_metacells_genes_linear_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_gene_log_linear_fraction_matrix" href="#Metacells.Contracts.metacell_gene_log_linear_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_log_linear_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_gene_log_linear_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The log base 2 of the linear fraction of the UMIs of each non-excluded gene in each metacell, out of the total UMIs. This adds some gene fraction regularization to deal with zero fractions. Using the log makes it easier to visualize, and the difference between log values (the &quot;fold factor&quot;, log base 2 of the ratio between the expression levels) is a good measure of difference between gene expression levels.
</p>
<p>This matrix is populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_genes_log_linear_fractions!">
<code>compute_metacells_genes_log_linear_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_gene_covered_fraction_matrix" href="#Metacells.Contracts.metacell_gene_covered_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_covered_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_gene_covered_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The linear fraction of the UMIs of each covered gene in each metacell, out of the total covered UMIs. By considering only the covered genes this avoid the impact of highly-expressed lateral genes (e.g., cell cycle).
</p>
<p>This matrix is populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_genes_covered_fractions!">
<code>compute_metacells_genes_covered_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_gene_log_covered_fraction_matrix" href="#Metacells.Contracts.metacell_gene_log_covered_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_log_covered_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_gene_log_covered_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The log base 2 of the linear fraction of the UMIs of each covered gene in each metacell, out of the total covered UMIs. This adds some gene fraction regularization to deal with zero fractions.
</p>
<p>This matrix is populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_genes_log_covered_fractions!">
<code>compute_metacells_genes_log_covered_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_gene_geomean_fraction_matrix" href="#Metacells.Contracts.metacell_gene_geomean_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_geomean_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_gene_geomean_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The geomean fraction of the UMIs of each gene in each metacell. We use geomean in an attempt to combat the disproportionate effect of a few cells with very high gene expression (&quot;bursty&quot; genes), and then normalizes the fractions to sum to one. While effective, this has the unfortunate effect of inflating the value of weak genes.
</p>
<p>TODO: Get rid of this.
</p>
<p>This matrix is populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_genes_geomean_fractions!">
<code>compute_metacells_genes_geomean_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_gene_log_geomean_fraction_matrix" href="#Metacells.Contracts.metacell_gene_log_geomean_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_log_geomean_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_gene_log_geomean_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The log base 2 of the geomean fraction of the UMIs of each gene in each metacell. This adds some gene fraction regularization to deal with zero fractions.
</p>
<p>TODO: Get rid of this.
</p>
<p>This matrix is populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_genes_log_geomean_fractions!">
<code>compute_metacells_genes_log_geomean_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<h3 id="Metacells-Distances">
<a class="docs-heading-anchor" href="#Metacells-Distances">Metacells Distances
</a>
<a id="Metacells-Distances-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Metacells-Distances" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_metacell_euclidean_skeleton_distance" href="#Metacells.Contracts.metacell_metacell_euclidean_skeleton_distance">
<code>Metacells.Contracts.metacell_metacell_euclidean_skeleton_distance
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_metacell_euclidean_skeleton_distance(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The Euclidean distance between the log of the covered fraction of the skeleton genes between the metacells.
</p>
<p>This matrix is computed by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_euclidean_distances!">
<code>compute_metacells_euclidean_distances!
</code>
</a>.
</p>
<p>TODO: Get rid of this?
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_metacell_max_skeleton_fold_distance" href="#Metacells.Contracts.metacell_metacell_max_skeleton_fold_distance">
<code>Metacells.Contracts.metacell_metacell_max_skeleton_fold_distance
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_metacell_max_skeleton_fold_distance(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The maximal significant fold factor between the covered fraction of skeleton genes between the metacells. This uses heuristics to require the fold factor be based on a sufficient number of UMIs to be robust.
</p>
<p>This matrix may be populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_max_skeleton_fold_distances!">
<code>compute_metacells_max_skeleton_fold_distances!
</code>
</a>.
</p>
<p>TODO: Get rid of this?
</p>
</div>
</section>
</article>
<h2 id="Blocks">
<a class="docs-heading-anchor" href="#Blocks">Blocks
</a>
<a id="Blocks-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Blocks" title="Permalink">
</a>
</h2>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_axis" href="#Metacells.Contracts.block_axis">
<code>Metacells.Contracts.block_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of blocks, which are distinct groups of metacells with &quot;very close&quot; cell state. The metacells in each block all have &quot;very close&quot; estimates of the covered fractions of the skeleton genes. Using blocks instead of metacells allows us to more uniformly sample the overall manifold. Highly sampled cell states would have blocks with larger number of higher-quality metacells, while sparsely sampled cell states would have blocks with fewer lower-quality metacells.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_block_vector" href="#Metacells.Contracts.metacell_block_vector">
<code>Metacells.Contracts.metacell_block_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_block_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The unique block each metacell belongs to.
</p>
<p>This vector is populated by 
<a href="compute_blocks.html#Metacells.ComputeBlocks.compute_blocks!">
<code>compute_blocks!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_n_metacells_vector" href="#Metacells.Contracts.block_n_metacells_vector">
<code>Metacells.Contracts.block_n_metacells_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_n_metacells_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The number of metacells in each block.
</p>
<p>This vector is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_n_metacells!">
<code>compute_blocks_n_metacells!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_n_cells_vector" href="#Metacells.Contracts.block_n_cells_vector">
<code>Metacells.Contracts.block_n_cells_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_n_cells_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The number of cells in the metacells in each block.
</p>
<p>This vector is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_n_cells!">
<code>compute_blocks_n_cells!
</code>
</a>.
</p>
</div>
</section>
</article>
<h3 id="Blocks-UMIs">
<a class="docs-heading-anchor" href="#Blocks-UMIs">Blocks UMIs
</a>
<a id="Blocks-UMIs-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Blocks-UMIs" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_UMIs_matrix" href="#Metacells.Contracts.block_gene_UMIs_matrix">
<code>Metacells.Contracts.block_gene_UMIs_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_UMIs_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of each gene in each block.
</p>
<p>This matrix is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_genes_UMIs!">
<code>compute_blocks_genes_UMIs!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_total_UMIs_vector" href="#Metacells.Contracts.block_total_UMIs_vector">
<code>Metacells.Contracts.block_total_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_total_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of all the non-excluded genes in each block. This is used as the denominator for 
<code>linear_fraction
</code>.
</p>
<p>This vector is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_total_UMIs!">
<code>compute_blocks_total_UMIs!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_covered_UMIs_vector" href="#Metacells.Contracts.block_covered_UMIs_vector">
<code>Metacells.Contracts.block_covered_UMIs_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_covered_UMIs_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of UMIs of all the covered genes in each block. This is used as the denominator for 
<code>covered_fraction
</code>.
</p>
<p>This vector is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_covered_UMIs!">
<code>compute_blocks_covered_UMIs!
</code>
</a>.
</p>
</div>
</section>
</article>
<h3 id="Blocks-Fractions">
<a class="docs-heading-anchor" href="#Blocks-Fractions">Blocks Fractions
</a>
<a id="Blocks-Fractions-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Blocks-Fractions" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_linear_fraction_matrix" href="#Metacells.Contracts.block_gene_linear_fraction_matrix">
<code>Metacells.Contracts.block_gene_linear_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_linear_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The linear fraction of the UMIs of each non-excluded gene out of the total UMIs in each block.
</p>
<p>This matrix is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_genes_linear_fractions!">
<code>compute_blocks_genes_linear_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_log_linear_fraction_matrix" href="#Metacells.Contracts.block_gene_log_linear_fraction_matrix">
<code>Metacells.Contracts.block_gene_log_linear_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_log_linear_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The log base 2 of the linear fraction of the UMIs of each non-excluded gene out of the total UMIs in each block. This adds some gene fraction regularization to deal with zero fractions.
</p>
<p>This matrix is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_genes_log_linear_fractions!">
<code>compute_blocks_genes_log_linear_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_covered_fraction_matrix" href="#Metacells.Contracts.block_gene_covered_fraction_matrix">
<code>Metacells.Contracts.block_gene_covered_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_covered_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The linear fraction of the UMIs of each covered gene out of the total covered UMIs in each block.
</p>
<p>This matrix is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_genes_covered_fractions!">
<code>compute_blocks_genes_covered_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_log_covered_fraction_matrix" href="#Metacells.Contracts.block_gene_log_covered_fraction_matrix">
<code>Metacells.Contracts.block_gene_log_covered_fraction_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_log_covered_fraction_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The log base 2 of the linear fraction of the UMIs of each covered gene out of the total covered UMIs in each block. This adds some gene fraction regularization to deal with zero fractions.
</p>
<p>This matrix is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_genes_log_covered_fractions!">
<code>compute_blocks_genes_log_covered_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<h3 id="Blocks-Distances">
<a class="docs-heading-anchor" href="#Blocks-Distances">Blocks Distances
</a>
<a id="Blocks-Distances-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Blocks-Distances" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_block_mean_euclidean_skeleton_distance" href="#Metacells.Contracts.block_block_mean_euclidean_skeleton_distance">
<code>Metacells.Contracts.block_block_mean_euclidean_skeleton_distance
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_block_mean_euclidean_skeleton_distance(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The mean Euclidean skeleton genes covered fractions distance between the metacells of the blocks.
</p>
<p>This matrix may be computed by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_mean_euclidean_distances!">
<code>compute_blocks_mean_euclidean_distances!
</code>
</a>.
</p>
<p>TODO: Get rid of this?
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_block_max_skeleton_fold_distance" href="#Metacells.Contracts.block_block_max_skeleton_fold_distance">
<code>Metacells.Contracts.block_block_max_skeleton_fold_distance
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_block_max_skeleton_fold_distance(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The maximal significant skeleton genes covered fractions fold factor between metacells of the blocks.
</p>
<p>This matrix may be populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_max_skeleton_fold_distances!">
<code>compute_blocks_max_skeleton_fold_distances!
</code>
</a>.
</p>
<p>TODO: Get rid of this?
</p>
</div>
</section>
</article>
<h3 id="Blocks-Neighborhoods">
<a class="docs-heading-anchor" href="#Blocks-Neighborhoods">Blocks Neighborhoods
</a>
<a id="Blocks-Neighborhoods-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Blocks-Neighborhoods" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_block_is_in_neighborhood_matrix" href="#Metacells.Contracts.block_block_is_in_neighborhood_matrix">
<code>Metacells.Contracts.block_block_is_in_neighborhood_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_block_is_in_neighborhood_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>For each block, the (column) mask of nearby blocks in its immediate neighborhood. The neighborhood consists of a small number of very close blocks.
</p>
<p>This matrix is populated by 
<a href="compute_blocks.html#Metacells.ComputeBlocks.compute_blocks_is_in_neighborhood!">
<code>compute_blocks_is_in_neighborhood!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_n_neighborhood_blocks_vector" href="#Metacells.Contracts.block_n_neighborhood_blocks_vector">
<code>Metacells.Contracts.block_n_neighborhood_blocks_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_n_neighborhood_blocks_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of blocks in the neighborhood centered at each block.
</p>
<p>This vector is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_n_neighborhood_blocks!">
<code>compute_blocks_n_neighborhood_blocks!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_n_neighborhood_metacells_vector" href="#Metacells.Contracts.block_n_neighborhood_metacells_vector">
<code>Metacells.Contracts.block_n_neighborhood_metacells_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_n_neighborhood_metacells_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of metacells in the blocks of the neighborhood centered at each block.
</p>
<p>This vector is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_n_neighborhood_metacells!">
<code>compute_blocks_n_neighborhood_metacells!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_n_neighborhood_cells_vector" href="#Metacells.Contracts.block_n_neighborhood_cells_vector">
<code>Metacells.Contracts.block_n_neighborhood_cells_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_n_neighborhood_cells_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of cells in the metacells in the blocks of the neighborhood centered at each block.
</p>
<p>This vector is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_n_neighborhood_cells!">
<code>compute_blocks_n_neighborhood_cells!
</code>
</a>.
</p>
</div>
</section>
</article>
<h3 id="Blocks-Environments">
<a class="docs-heading-anchor" href="#Blocks-Environments">Blocks Environments
</a>
<a id="Blocks-Environments-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Blocks-Environments" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_block_is_in_environment_matrix" href="#Metacells.Contracts.block_block_is_in_environment_matrix">
<code>Metacells.Contracts.block_block_is_in_environment_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_block_is_in_environment_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>For each block, the (column) mask of nearby blocks in its expanded environment. The environment consists of a larger region than the neighborhood, which is still close to the center block.
</p>
<p>This matrix is populated by 
<a href="compute_blocks.html#Metacells.ComputeBlocks.compute_blocks_is_in_environment!">
<code>compute_blocks_is_in_environment!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_n_environment_blocks_vector" href="#Metacells.Contracts.block_n_environment_blocks_vector">
<code>Metacells.Contracts.block_n_environment_blocks_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_n_environment_blocks_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of blocks in the environment centered at each block.
</p>
<p>This vector is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_n_environment_blocks!">
<code>compute_blocks_n_environment_blocks!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_n_environment_metacells_vector" href="#Metacells.Contracts.block_n_environment_metacells_vector">
<code>Metacells.Contracts.block_n_environment_metacells_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_n_environment_metacells_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of metacells in the blocks of the environment centered at each block.
</p>
<p>This vector is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_n_environment_metacells!">
<code>compute_blocks_n_environment_metacells!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_n_environment_cells_vector" href="#Metacells.Contracts.block_n_environment_cells_vector">
<code>Metacells.Contracts.block_n_environment_cells_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_n_environment_cells_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total number of cells in the metacells in the blocks of the environment centered at each block.
</p>
<p>This vector is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_n_environment_cells!">
<code>compute_blocks_n_environment_cells!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_is_environment_marker_matrix" href="#Metacells.Contracts.block_gene_is_environment_marker_matrix">
<code>Metacells.Contracts.block_gene_is_environment_marker_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_is_environment_marker_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>A mask of covered genes that distinguish between cell states in each environment.
</p>
<p>This matrix is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_genes_is_environment_markers!">
<code>compute_blocks_genes_is_environment_markers!
</code>
</a>.
</p>
</div>
</section>
</article>
<h3 id="Blocks-Correlations">
<a class="docs-heading-anchor" href="#Blocks-Correlations">Blocks Correlations
</a>
<a id="Blocks-Correlations-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Blocks-Correlations" title="Permalink">
</a>
</h3>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_neighborhood_correlation_matrix" href="#Metacells.Contracts.block_gene_neighborhood_correlation_matrix">
<code>Metacells.Contracts.block_gene_neighborhood_correlation_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_neighborhood_correlation_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The correlation between cells and metacells gene expression levels in each block&#39;s neighborhood. This is zero for excluded genes.
</p>
<p>This matrix is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_genes_neighborhood_correlations!">
<code>compute_blocks_genes_neighborhood_correlations!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_is_correlated_with_metacells" href="#Metacells.Contracts.block_gene_is_correlated_with_metacells">
<code>Metacells.Contracts.block_gene_is_correlated_with_metacells
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_is_correlated_with_metacells(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>Whether each gene is strongly correlated between cells and metacells in each neighborhood.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_mean_neighborhood_correlation_vector" href="#Metacells.Contracts.block_mean_neighborhood_correlation_vector">
<code>Metacells.Contracts.block_mean_neighborhood_correlation_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_mean_neighborhood_correlation_vector(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The mean correlation between cells and metacells gene expression levels in each block&#39;s neighborhood.
</p>
<p>This matrix is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_mean_neighborhood_correlations!">
<code>compute_blocks_mean_neighborhood_correlations!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_mean_covered_neighborhood_correlation_vector" href="#Metacells.Contracts.block_mean_covered_neighborhood_correlation_vector">
<code>Metacells.Contracts.block_mean_covered_neighborhood_correlation_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_mean_neighborhood_covered_correlation_vector(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The mean correlation between cells and metacells covered gene expression levels in each block&#39;s neighborhood.
</p>
<p>This matrix is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_mean_covered_neighborhood_correlations!">
<code>compute_blocks_mean_covered_neighborhood_correlations!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_mean_modules_neighborhood_correlation_vector" href="#Metacells.Contracts.block_mean_modules_neighborhood_correlation_vector">
<code>Metacells.Contracts.block_mean_modules_neighborhood_correlation_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_mean_neighborhood_modules_correlation_vector(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The mean correlation between cells and metacells modules gene expression levels in each block&#39;s neighborhood.
</p>
<p>This matrix is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_mean_modules_neighborhood_correlations!">
<code>compute_blocks_mean_modules_neighborhood_correlations!
</code>
</a>.
</p>
</div>
</section>
</article>
<h2 id="Type-Annotations">
<a class="docs-heading-anchor" href="#Type-Annotations">Type Annotations
</a>
<a id="Type-Annotations-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Type-Annotations" title="Permalink">
</a>
</h2>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.type_axis" href="#Metacells.Contracts.type_axis">
<code>Metacells.Contracts.type_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">type_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of types, which are distinct named biological cell states. Types are convenient labels manually assigned to large groups of cells, metacells, blocks, etc. Types are not associated with an exact biological cell state, but rather with a set of related biological cell states, possibly along a gradient of such states. The resolution of the type labels therefore depends on the data set and the type of analysis.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.type_color_vector" href="#Metacells.Contracts.type_color_vector">
<code>Metacells.Contracts.type_color_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">type_color_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>A unique color for each type, for visualizations.
</p>
<p>This vector is created in a supervised way based on biological considerations and conventions (e.g., red blood cells are often given some red color). It is also possible to use 
<code>Chameleon
</code> to automatically assign colors to cell types based on some gene expression levels.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.metacell_type_vector" href="#Metacells.Contracts.metacell_type_vector">
<code>Metacells.Contracts.metacell_type_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">metacell_type_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The type each metacell belongs to. This can be assigned in many ways, from manual annotation to using projection methods on an annotated atlas. Often there are multiple type annotations due to different methods. By convention these alternate type vectors are named 
<code>type.something
</code> (e.g., 
<code>type.projected
</code>, 
<code>type.manual
</code>) and the simple 
<code>type
</code> is used for &quot;the&quot; type of each metacell.
</p>
<p>This vector is created in a supervised way based on biological considerations. Alternatively, if one has a cell type vector from somewhere, it can be populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_metacells_types_by_cells!">
<code>compute_metacells_types_by_cells!
</code>
</a> which assigns to each metacell the most frequent type of the cells it contains.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.cell_type_vector" href="#Metacells.Contracts.cell_type_vector">
<code>Metacells.Contracts.cell_type_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">cell_type_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The type each cell belongs to. By convention 
<code>type
</code> is used for &quot;the&quot; type of the cell, which is compatible with the type of the metacell it belongs to, and other 
<code>type.something
</code> are used for types computed in other ways, such as importing data from a type-annotated single-cell atlas, or by running some classifier directly on the cells.
</p>
<p>This vector can be populated by 
<a href="analyze_metacells.html#Metacells.AnalyzeMetacells.compute_cells_types_by_metacells!">
<code>compute_cells_types_by_metacells!
</code>
</a> which assigns to each cell the type of the metacell it belongs to.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_type_vector" href="#Metacells.Contracts.block_type_vector">
<code>Metacells.Contracts.block_type_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_type_vector(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The type each block belongs to. This is typically taken to be the most frequent metacell type in the block.
</p>
<p>This vector is populated by 
<a href="analyze_blocks.html#Metacells.AnalyzeBlocks.compute_blocks_types!">
<code>compute_blocks_types!
</code>
</a> based on the metacell types in each block.
</p>
</div>
</section>
</article>
<h2 id="Modules">
<a class="docs-heading-anchor" href="#Modules">Modules
</a>
<a id="Modules-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Modules" title="Permalink">
</a>
</h2>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.module_axis" href="#Metacells.Contracts.module_axis">
<code>Metacells.Contracts.module_axis
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">module_axis(expectation::ContractExpectation)::Pair{AxisKey, AxisSpecification}
</code>
</pre>
<p>The axis of modules, which are groups of genes that together can predict the rest of the covered genes. Each module is based on a single anchor regulator gene, and may contain additional regulator and regular genes. The set and composition of modules varies between blocks across the manifold. The names of the modules are the names of their anchors (for interpretability) followed by a 
<code>.MOD
</code> suffix (to clarify this is the name of a module and not a gene).
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.module_anchor_gene_vector" href="#Metacells.Contracts.module_anchor_gene_vector">
<code>Metacells.Contracts.module_anchor_gene_vector
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">module_anchor_gene_vector(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The anchor gene each module is based on. Even though the &quot;same&quot; module in different blocks is based on the same anchor gene, the rest of the composition of the module may vary wildly, even between neighboring blocks.
</p>
<p>This matrix is populated by 
<a href="compute_modules.html#Metacells.ComputeModules.compute_blocks_modules!">
<code>compute_blocks_modules!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_gene_module_matrix" href="#Metacells.Contracts.block_gene_module_matrix">
<code>Metacells.Contracts.block_gene_module_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_gene_module_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The module each gene belongs to in each block. Most genes do not belong to any module and therefore have an empty string as the value.
</p>
<p>This matrix is populated by 
<a href="compute_modules.html#Metacells.ComputeModules.compute_blocks_modules!">
<code>compute_blocks_modules!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_module_is_found_matrix" href="#Metacells.Contracts.block_module_is_found_matrix">
<code>Metacells.Contracts.block_module_is_found_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_module_is_found_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>A mask of the modules that were found for each block. Due to 
<code>Daf
</code> limitations, the modules axis must cover all the modules of all the blocks. This mask specifies which of the modules were actually found for each of the blocks.
</p>
<p>This matrix is populated by 
<a href="compute_modules.html#Metacells.ComputeModules.compute_blocks_modules!">
<code>compute_blocks_modules!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_module_is_strong_matrix" href="#Metacells.Contracts.block_module_is_strong_matrix">
<code>Metacells.Contracts.block_module_is_strong_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_module_is_strong_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>A mask of the strong modules that have enough UMIs in enough cells. Since cells have much less UMIs than metacells, not all modules have a high enough number of UMIs in enough cells to be useful.
</p>
<p>This matrix is populated by 
<a href="analyze_modules.html#Metacells.AnalyzeModules.compute_blocks_modules_is_strong!">
<code>compute_blocks_modules_is_strong!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_module_n_genes_matrix" href="#Metacells.Contracts.block_module_n_genes_matrix">
<code>Metacells.Contracts.block_module_n_genes_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_module_n_genes_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The number of genes in each gene module in each block. This is zero for the extra (not-found) gene modules.
</p>
<p>This matrix is populated by 
<a href="analyze_modules.html#Metacells.AnalyzeModules.compute_blocks_modules_n_genes!">
<code>compute_blocks_modules_n_genes!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_module_n_skeletons_matrix" href="#Metacells.Contracts.block_module_n_skeletons_matrix">
<code>Metacells.Contracts.block_module_n_skeletons_matrix
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_module_n_skeletons_matrix(expectation::ContractExpectation)::Pair{MatrixKey, DataSpecification}
</code>
</pre>
<p>The number of skeleton genes in each gene module in each block. This is zero for the extra (not-found) gene modules.
</p>
<p>This matrix is populated by 
<a href="analyze_modules.html#Metacells.AnalyzeModules.compute_blocks_modules_n_skeletons!">
<code>compute_blocks_modules_n_skeletons!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_metacell_module_environment_total_UMIs_tensor" href="#Metacells.Contracts.block_metacell_module_environment_total_UMIs_tensor">
<code>Metacells.Contracts.block_metacell_module_environment_total_UMIs_tensor
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_metacell_module_environment_total_UMIs_tensor(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The total UMIs of each environment module in each environment metacell. This is zero for metacells that are not part of the block&#39;s environment, or modules that aren&#39;t found in that environment.
</p>
<p>This vector is populated by 
<a href="analyze_modules.html#Metacells.AnalyzeModules.compute_block_metacell_module_environment_total_UMIs!">
<code>compute_block_metacell_module_environment_total_UMIs!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_metacell_module_environment_linear_fraction_tensor" href="#Metacells.Contracts.block_metacell_module_environment_linear_fraction_tensor">
<code>Metacells.Contracts.block_metacell_module_environment_linear_fraction_tensor
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_metacell_module_environment_linear_fraction_tensor(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The linear fraction of the total UMIs of each environment module in each environment metacell, out of the total UMIs. This is zero for metacells that are not part of the block&#39;s environment, or modules that aren&#39;t found in that environment.
</p>
<p>This vector is populated by 
<a href="analyze_modules.html#Metacells.AnalyzeModules.compute_block_metacell_module_environment_linear_fractions!">
<code>compute_block_metacell_module_environment_linear_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_metacell_module_environment_log_linear_fraction_tensor" href="#Metacells.Contracts.block_metacell_module_environment_log_linear_fraction_tensor">
<code>Metacells.Contracts.block_metacell_module_environment_log_linear_fraction_tensor
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_metacell_module_environment_log_linear_fraction_tensor(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The log base 2 of the linear fraction of the total UMIs of each environment module in each environment metacell, out of the total UMIs. This adds some gene fraction regularization to deal with zero fractions. This is zero for metacells that are not part of the block&#39;s environment, or modules that aren&#39;t found in that environment.
</p>
<p>This vector is populated by 
<a href="analyze_modules.html#Metacells.AnalyzeModules.compute_block_metacell_module_environment_log_linear_fractions!">
<code>compute_block_metacell_module_environment_log_linear_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_metacell_module_environment_covered_fraction_tensor" href="#Metacells.Contracts.block_metacell_module_environment_covered_fraction_tensor">
<code>Metacells.Contracts.block_metacell_module_environment_covered_fraction_tensor
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_metacell_module_environment_covered_fraction_tensor(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The linear fraction of the total UMIs of each environment module in each environment metacell, out of the total covered UMIs. This is zero for metacells that are not part of the block&#39;s environment, or modules that aren&#39;t found in that environment.
</p>
<p>This vector is populated by 
<a href="analyze_modules.html#Metacells.AnalyzeModules.compute_block_metacell_module_environment_covered_fractions!">
<code>compute_block_metacell_module_environment_covered_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Contracts.block_metacell_module_environment_log_covered_fraction_tensor" href="#Metacells.Contracts.block_metacell_module_environment_log_covered_fraction_tensor">
<code>Metacells.Contracts.block_metacell_module_environment_log_covered_fraction_tensor
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">block_metacell_module_environment_log_covered_fraction_tensor(expectation::ContractExpectation)::Pair{VectorKey, DataSpecification}
</code>
</pre>
<p>The log base 2 of the linear fraction of the total UMIs of each environment module in each environment metacell, out of the total covered UMIs. This adds some gene fraction regularization to deal with zero fractions. This is zero for metacells that are not part of the block&#39;s environment, or modules that aren&#39;t found in that environment.
</p>
<p>This vector is populated by 
<a href="analyze_modules.html#Metacells.AnalyzeModules.compute_block_metacell_module_environment_log_covered_fractions!">
<code>compute_block_metacell_module_environment_log_covered_fractions!
</code>
</a>.
</p>
</div>
</section>
</article>
<h2 id="Index">
<a class="docs-heading-anchor" href="#Index">Index
</a>
<a id="Index-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Index" title="Permalink">
</a>
</h2>
<ul>
<li>
<a href="contracts.html#Metacells.Contracts">
<code>Metacells.Contracts
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_axis">
<code>Metacells.Contracts.block_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_block_is_in_environment_matrix">
<code>Metacells.Contracts.block_block_is_in_environment_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_block_is_in_neighborhood_matrix">
<code>Metacells.Contracts.block_block_is_in_neighborhood_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_block_max_skeleton_fold_distance">
<code>Metacells.Contracts.block_block_max_skeleton_fold_distance
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_block_mean_euclidean_skeleton_distance">
<code>Metacells.Contracts.block_block_mean_euclidean_skeleton_distance
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_covered_UMIs_vector">
<code>Metacells.Contracts.block_covered_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_UMIs_matrix">
<code>Metacells.Contracts.block_gene_UMIs_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_covered_fraction_matrix">
<code>Metacells.Contracts.block_gene_covered_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_is_correlated_with_metacells">
<code>Metacells.Contracts.block_gene_is_correlated_with_metacells
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_is_environment_marker_matrix">
<code>Metacells.Contracts.block_gene_is_environment_marker_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_linear_fraction_matrix">
<code>Metacells.Contracts.block_gene_linear_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_log_covered_fraction_matrix">
<code>Metacells.Contracts.block_gene_log_covered_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_log_linear_fraction_matrix">
<code>Metacells.Contracts.block_gene_log_linear_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_module_matrix">
<code>Metacells.Contracts.block_gene_module_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_gene_neighborhood_correlation_matrix">
<code>Metacells.Contracts.block_gene_neighborhood_correlation_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_mean_covered_neighborhood_correlation_vector">
<code>Metacells.Contracts.block_mean_covered_neighborhood_correlation_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_mean_modules_neighborhood_correlation_vector">
<code>Metacells.Contracts.block_mean_modules_neighborhood_correlation_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_mean_neighborhood_correlation_vector">
<code>Metacells.Contracts.block_mean_neighborhood_correlation_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_metacell_module_environment_covered_fraction_tensor">
<code>Metacells.Contracts.block_metacell_module_environment_covered_fraction_tensor
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_metacell_module_environment_linear_fraction_tensor">
<code>Metacells.Contracts.block_metacell_module_environment_linear_fraction_tensor
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_metacell_module_environment_log_covered_fraction_tensor">
<code>Metacells.Contracts.block_metacell_module_environment_log_covered_fraction_tensor
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_metacell_module_environment_log_linear_fraction_tensor">
<code>Metacells.Contracts.block_metacell_module_environment_log_linear_fraction_tensor
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_metacell_module_environment_total_UMIs_tensor">
<code>Metacells.Contracts.block_metacell_module_environment_total_UMIs_tensor
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_module_is_found_matrix">
<code>Metacells.Contracts.block_module_is_found_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_module_is_strong_matrix">
<code>Metacells.Contracts.block_module_is_strong_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_module_n_genes_matrix">
<code>Metacells.Contracts.block_module_n_genes_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_module_n_skeletons_matrix">
<code>Metacells.Contracts.block_module_n_skeletons_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_n_cells_vector">
<code>Metacells.Contracts.block_n_cells_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_n_environment_blocks_vector">
<code>Metacells.Contracts.block_n_environment_blocks_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_n_environment_cells_vector">
<code>Metacells.Contracts.block_n_environment_cells_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_n_environment_metacells_vector">
<code>Metacells.Contracts.block_n_environment_metacells_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_n_metacells_vector">
<code>Metacells.Contracts.block_n_metacells_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_n_neighborhood_blocks_vector">
<code>Metacells.Contracts.block_n_neighborhood_blocks_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_n_neighborhood_cells_vector">
<code>Metacells.Contracts.block_n_neighborhood_cells_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_n_neighborhood_metacells_vector">
<code>Metacells.Contracts.block_n_neighborhood_metacells_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_total_UMIs_vector">
<code>Metacells.Contracts.block_total_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.block_type_vector">
<code>Metacells.Contracts.block_type_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_axis">
<code>Metacells.Contracts.cell_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_covered_UMIs_vector">
<code>Metacells.Contracts.cell_covered_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_excluded_UMIs_vector">
<code>Metacells.Contracts.cell_excluded_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_gene_UMIs_matrix">
<code>Metacells.Contracts.cell_gene_UMIs_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_is_excluded_vector">
<code>Metacells.Contracts.cell_is_excluded_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_metacell_vector">
<code>Metacells.Contracts.cell_metacell_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_mitochondrial_UMIs_vector">
<code>Metacells.Contracts.cell_mitochondrial_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_ribosomal_UMIs_vector">
<code>Metacells.Contracts.cell_ribosomal_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_total_UMIs_vector">
<code>Metacells.Contracts.cell_total_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.cell_type_vector">
<code>Metacells.Contracts.cell_type_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_axis">
<code>Metacells.Contracts.gene_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_covered_vector">
<code>Metacells.Contracts.gene_is_covered_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_excluded_vector">
<code>Metacells.Contracts.gene_is_excluded_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_forbidden_vector">
<code>Metacells.Contracts.gene_is_forbidden_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_lateral_vector">
<code>Metacells.Contracts.gene_is_lateral_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_marker_vector">
<code>Metacells.Contracts.gene_is_marker_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_mitochondrial_vector">
<code>Metacells.Contracts.gene_is_mitochondrial_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_regulator_vector">
<code>Metacells.Contracts.gene_is_regulator_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_ribosomal_vector">
<code>Metacells.Contracts.gene_is_ribosomal_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_is_skeleton_vector">
<code>Metacells.Contracts.gene_is_skeleton_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.gene_marker_rank_vector">
<code>Metacells.Contracts.gene_marker_rank_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_axis">
<code>Metacells.Contracts.metacell_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_block_vector">
<code>Metacells.Contracts.metacell_block_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_covered_UMIs_vector">
<code>Metacells.Contracts.metacell_covered_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_gene_UMIs_matrix">
<code>Metacells.Contracts.metacell_gene_UMIs_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_gene_covered_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_covered_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_gene_geomean_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_geomean_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_gene_linear_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_linear_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_gene_log_covered_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_log_covered_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_gene_log_geomean_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_log_geomean_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_gene_log_linear_fraction_matrix">
<code>Metacells.Contracts.metacell_gene_log_linear_fraction_matrix
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_metacell_euclidean_skeleton_distance">
<code>Metacells.Contracts.metacell_metacell_euclidean_skeleton_distance
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_metacell_max_skeleton_fold_distance">
<code>Metacells.Contracts.metacell_metacell_max_skeleton_fold_distance
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_n_cells_vector">
<code>Metacells.Contracts.metacell_n_cells_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_total_UMIs_vector">
<code>Metacells.Contracts.metacell_total_UMIs_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.metacell_type_vector">
<code>Metacells.Contracts.metacell_type_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.module_anchor_gene_vector">
<code>Metacells.Contracts.module_anchor_gene_vector
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.module_axis">
<code>Metacells.Contracts.module_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.type_axis">
<code>Metacells.Contracts.type_axis
</code>
</a>
</li>
<li>
<a href="contracts.html#Metacells.Contracts.type_color_vector">
<code>Metacells.Contracts.type_color_vector
</code>
</a>
</li>
</ul>
</article>
<nav class="docs-footer">
<a class="docs-footer-prevpage" href="gmara.html">« Gmara
</a>
<a class="docs-footer-nextpage" href="analyze_cells.html">Analyze Cells »
</a>
<div class="flexbox-break">
</div>
<p class="footer-message">Powered by 
<a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl
</a> and the 
<a href="https://julialang.org/">Julia Programming Language
</a>.
</p>
</nav>
</div>
<div class="modal" id="documenter-settings">
<div class="modal-background">
</div>
<div class="modal-card">
<header class="modal-card-head">
<p class="modal-card-title">Settings
</p>
<button class="delete">
</button>
</header>
<section class="modal-card-body">
<p>
<label class="label">Theme
</label>
<div class="select">
<select id="documenter-themepicker">
<option value="auto">Automatic (OS)
</option>
<option value="documenter-light">documenter-light
</option>
<option value="documenter-dark">documenter-dark
</option>
<option value="catppuccin-latte">catppuccin-latte
</option>
<option value="catppuccin-frappe">catppuccin-frappe
</option>
<option value="catppuccin-macchiato">catppuccin-macchiato
</option>
<option value="catppuccin-mocha">catppuccin-mocha
</option>
</select>
</div>
</p>
<hr/>
<p>This document was generated with 
<a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl
</a> version 1.12.0. Using Julia version 1.11.5.
</p>
</section>
<footer class="modal-card-foot">
</footer>
</div>
</div>
</div>
</body>
</html>
