
<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="UTF-8"/>
<meta name="viewport" content="width=device-width, initial-scale=1.0"/>
<title>Programs · Metacells.jl v0.1.0
</title>
<meta name="title" content="Programs · Metacells.jl v0.1.0"/>
<meta property="og:title" content="Programs · Metacells.jl v0.1.0"/>
<meta property="twitter:title" content="Programs · Metacells.jl v0.1.0"/>
<meta name="description" content="Documentation for Metacells.jl v0.1.0."/>
<meta property="og:description" content="Documentation for Metacells.jl v0.1.0."/>
<meta property="twitter:description" content="Documentation for Metacells.jl v0.1.0."/>
<script data-outdated-warner src="assets/warner.js">
</script>
<link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/>
<script>documenterBaseURL="."
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js">
</script>
<script src="search_index.js">
</script>
<script src="siteinfo.js">
</script>
<script src="../versions.js">
</script>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/>
<link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/>
<script src="assets/themeswap.js">
</script>
</head>
<body>
<div id="documenter">
<nav class="docs-sidebar">
<div class="docs-package-name">
<span class="docs-autofit">
<a href="index.html">Metacells.jl v0.1.0
</a>
</span>
</div>
<button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)
</button>
<ul class="docs-menu">
<li>
<a class="tocitem" href="index.html">Metacells
</a>
</li>
<li>
<a class="tocitem" href="gmara.html">Gmara
</a>
</li>
<li>
<a class="tocitem" href="contracts.html">Contracts
</a>
</li>
<li>
<a class="tocitem" href="identify_genes.html">Identify Genes
</a>
</li>
<li class="is-active">
<a class="tocitem" href="programs.html">Programs
</a>
<ul class="internal">
<li>
<a class="tocitem" href="#Index">
<span>Index
</span>
</a>
</li>
</ul>
</li>
<li>
<a class="tocitem" href="anndata_format.html">AnnData Format
</a>
</li>
<li>
<a class="tocitem" href="defaults.html">Defaults
</a>
</li>
</ul>
<div class="docs-version-selector field has-addons">
<div class="control">
<span class="docs-label button is-static is-size-7">Version
</span>
</div>
<div class="docs-selector control is-expanded">
<div class="select is-fullwidth is-size-7">
<select id="documenter-version-selector">
</select>
</div>
</div>
</div>
</nav>
<div class="docs-main">
<header class="docs-navbar">
<a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#">
</a>
<nav class="breadcrumb">
<ul class="is-hidden-mobile">
<li class="is-active">
<a href="programs.html">Programs
</a>
</li>
</ul>
<ul class="is-hidden-tablet">
<li class="is-active">
<a href="programs.html">Programs
</a>
</li>
</ul>
</nav>
<div class="docs-right">
<a class="docs-navbar-link" href="https://github.com/tanaylab/Metacells.jl/blob/main{path}?plain=1#L{line}" title="View the repository on GitHub">
<span class="docs-icon fa-brands">
</span>
<span class="docs-label is-hidden-touch">GitHub
</span>
</a>
<a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings">
</a>
<a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings">
</a>
</div>
</header>
<article class="content" id="documenter-page">
<h1 id="Programs">
<a class="docs-heading-anchor" href="#Programs">Programs
</a>
<a id="Programs-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Programs" title="Permalink">
</a>
</h1>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Programs" href="#Metacells.Programs">
<code>Metacells.Programs
</code>
</a> — 
<span class="docstring-category">Module
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<p>Approximate the manifold of actual cell states (captured by metacells) using linear programs in each local region.
</p>
</div>
</section>
</article>
<article class="docstring">
<header>
<a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring">
</a>
<a class="docstring-binding" id="Metacells.Programs.compute_blocks!" href="#Metacells.Programs.compute_blocks!">
<code>Metacells.Programs.compute_blocks!
</code>
</a> — 
<span class="docstring-category">Function
</span>
<span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring">
</span>
</header>
<section>
<div>
<pre>
<code class="language-julia hljs">function compute_blocks!(
    daf::DafWriter;
    gene_fraction_regularization::AbstractFloat = 2.0e-5,
    min_significant_gene_UMIs::Integer = 40,
    fold_confidence::AbstractFloat = 0.9,
    max_block_span::Real = 1,
    max_principal_components = 40,
    min_blocks_in_neighborhood::Integer = 4,
    min_metacells_in_neighborhood::Integer = 50,
    min_blocks_in_approximate_environment::Integer = 20,
    min_metacells_in_approximate_environment::Integer = 500,
    cross_validation_parts::Integer = 5,
    rng::AbstractRNG = default_rng(),
    overwrite::Bool = false,
)::Nothing
</code>
</pre>
<p>Group the metacells into blocks where the metacells in each one are &quot;similar&quot; in all the transcription factors. That is, each block is an approximation of a single cell state, assuming the transcription factors control the expression of the rest of the genes. Then, group these blocks into overlapping small tight neighborhoods, and larger environments, such that each block&#39;s environment can be reasonably approximated using a linear model (which is not computed), evaluated using cross-validation on the RMSE in the neighborhood at its core.
</p>
<p>First, we compute metacell-metacell distances based on the maximal fold factor between 
<code>is_transcription_factor
</code> genes, which aren&#39;t 
<code>is_forbidden_factor
</code>. The fold factor is log (base 2) of the gene expression using the 
<code>gene_fraction_regularization
</code>. For computing this fold factor, we ignore genes whose total UMIs in the compared metacells isn&#39;t at least 
<code>min_significant_gene_UMIs
</code>. We also we reduce the distance using the 
<code>fold_confidence
</code> based on the number of UMIs used to estimate the expression in the metacells, and the 
<code>divergence
</code> of the gene. Two metacells can only belong to the same block if the final fold factor is at most 
<code>max_block_span
</code> in all the genes.
</p>
<p>We then compute block-block distance which is the mean distance between the metacells of the blocks. Using this distances, we define a tight neighborhood of each block containing at least 
<code>min_blocks_in_neighborhood
</code> and 
<code>min_metacells_in_neighborhood
</code>. We also compute an approximate environment of at least 
<code>min_blocks_in_approximate_environment
</code> and 
<code>min_metacells_in_approximate_environment
</code>. Using this approximate environment, we compute 
<code>max_principal_components
</code> for the transcription factor genes, and figure out how many of them are useful to predict the value of the rest of the genes using 
<code>cross_validation_parts
</code> minimizing the RMSE in the neighborhood.
</p>
<p>Having determined the dimensionality of the environment of the block, we search for the optimal environment such that the RMSE of the cross validation of a linear model with this humber of principal components (computed on the transcription factor genes) for predicting the value of the rest of the genes.
</p>
<p>
<strong>Inputs
</strong>
</p>
<p>
<strong>Axes
</strong>
</p>
<p>
<strong>gene
</strong> (required): Sequenced genes.
</p>
<p>
<strong>metacell
</strong> (required): Minimal-sized groups of cells for robust point estimates.
</p>
<p>
<strong>Vectors
</strong>
</p>
<p>
<strong>metacell @ total_UMIs
</strong>::Union{UInt16, UInt32, UInt64, UInt8} (required): The total number of UMIs used to estimate the fraction of all the genes in each metacell.
</p>
<p>
<strong>gene @ divergence
</strong>::Union{Float32, Float64} (required): Scale fold factors of each gene by multiplying with (1 - divergence) of the gene.
</p>
<p>
<strong>gene @ is
<em>transcription
</em>factor
</strong>::Bool (required): A mask of genes that bind to the DNA.
</p>
<p>
<strong>gene @ is
<em>forbidden
</em>factor
</strong>::Bool (optional): A mask of genes that are forbidden from being used as predictive factors.
</p>
<p>
<strong>Matrices
</strong>
</p>
<p>
<strong>gene, metacell @ fraction
</strong>::Union{Float32, Float64} (required): The estimated fraction of the UMIs of each gene in each metacell.
</p>
<p>
<strong>gene, metacell @ total_UMIs
</strong>::Union{UInt16, UInt32, UInt64, UInt8} (required): The total number of UMIs used to estimate the fraction of each gene in each metacell.
</p>
<p>Additional inputs may be used depending on the parameter(s).
</p>
<p>
<strong>Outputs
</strong>
</p>
<p>
<strong>Axes
</strong>
</p>
<p>
<strong>block
</strong> (guaranteed): Distinct groups of metacells with &quot;very close&quot; estimated cell state.
</p>
<p>
<strong>Vectors
</strong>
</p>
<p>
<strong>metacell @ block
</strong>::AbstractString (guaranteed): The unique block each metacell belongs to.
</p>
<p>
<strong>Matrices
</strong>
</p>
<p>
<strong>block, block @ is
<em>in
</em>neighborhood
</strong>::Bool (guaranteed): For each block, the mask of nearby blocks in its immediate neighborhood.
</p>
<p>
<strong>block, block @ is
<em>in
</em>environment
</strong>::Bool (guaranteed): For each block, the mask of nearby blocks in its linear environment.
</p>
<p>Additional outputs may be created depending on the parameter(s).
</p>
</div>
</section>
</article>
<h2 id="Index">
<a class="docs-heading-anchor" href="#Index">Index
</a>
<a id="Index-1">
</a>
<a class="docs-heading-anchor-permalink" href="#Index" title="Permalink">
</a>
</h2>
<ul>
<li>
<a href="programs.html#Metacells.Programs">
<code>Metacells.Programs
</code>
</a>
</li>
<li>
<a href="programs.html#Metacells.Programs.compute_blocks!">
<code>Metacells.Programs.compute_blocks!
</code>
</a>
</li>
</ul>
</article>
<nav class="docs-footer">
<a class="docs-footer-prevpage" href="identify_genes.html">« Identify Genes
</a>
<a class="docs-footer-nextpage" href="anndata_format.html">AnnData Format »
</a>
<div class="flexbox-break">
</div>
<p class="footer-message">Powered by 
<a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl
</a> and the 
<a href="https://julialang.org/">Julia Programming Language
</a>.
</p>
</nav>
</div>
<div class="modal" id="documenter-settings">
<div class="modal-background">
</div>
<div class="modal-card">
<header class="modal-card-head">
<p class="modal-card-title">Settings
</p>
<button class="delete">
</button>
</header>
<section class="modal-card-body">
<p>
<label class="label">Theme
</label>
<div class="select">
<select id="documenter-themepicker">
<option value="auto">Automatic (OS)
</option>
<option value="documenter-light">documenter-light
</option>
<option value="documenter-dark">documenter-dark
</option>
<option value="catppuccin-latte">catppuccin-latte
</option>
<option value="catppuccin-frappe">catppuccin-frappe
</option>
<option value="catppuccin-macchiato">catppuccin-macchiato
</option>
<option value="catppuccin-mocha">catppuccin-mocha
</option>
</select>
</div>
</p>
<hr/>
<p>This document was generated with 
<a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl
</a> version 1.7.0. Using Julia version 1.10.5.
</p>
</section>
<footer class="modal-card-foot">
</footer>
</div>
</div>
</div>
</body>
</html>
